File: src/kyu_4/parseint_reloaded/python/parseint_reloaded.py
def parse_int(string):

    value = 0

    string = string.lower().replace(' and', '') # normalize

    millions, thousands, ones =  parse_periods(string)

    if millions:
        value += 1_000_000 * to_integer(millions)

    if thousands:
        value += 1000 * to_integer(thousands)

    if ones:
        value += to_integer(ones)

    return value

def parse_periods(text):
    """Break text into millions, thousands, and ones periods"""

    millions = ''
    thousands = ''
    ones = ''
    head = text

    if 'million' in head:
        millions, sep, head = head.partition('million')

    if 'thousand' in head:
        thousands, sep, head = head.partition('thousand')

    if head:
        ones = head

    return (millions.strip(), thousands.strip(), ones.strip())

def to_integer(text):
    """Convert text less than 'one thousand' to integer"""

    to_num = {
        'zero' : 0,
        'one' : 1,
        'two' : 2,
        'three' : 3,
        'four' : 4,
        'five' : 5,
        'six' : 6,
        'seven' : 7,
        'eight' : 8,
        'nine' : 9,
        'ten' : 10,
        'eleven' : 11,
        'twelve' : 12,
        'thirteen' : 13,
        'fourteen' : 14,
        'fifteen' : 15,
        'sixteen' : 16,
        'seventeen' : 17,
        'eighteen' : 18,
        'nineteen' : 19,
        'twenty' : 20,
        'thirty' : 30,
        'forty' : 40,
        'fifty' : 50,
        'sixty' : 60,
        'seventy' : 70,
        'eighty' : 80,
        'ninety' : 90
    }

    if not text:
        raise ValueError

    head = text
    result = 0

    if 'hundred' in head:
        hundreds, sep, head = head.partition('hundred')
        result += 100 * to_num[hundreds.strip()]

    if '-' in head:
        tens, sep, head = head.partition('-')
        result += to_num[tens.strip()]

    if head.strip():
        result += to_num[head.strip()]

    return result

File: src/kyu_4/parseint_reloaded/csharp/parseint_reloaded.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
public class Parser
{
    public static int ParseInt(string s)
    {
      int result = 0;
      int thousands = 0;
      int millions = 0;

      s = s.Replace('-', ' ');

      foreach (string token in s.Split())
      {
        switch (token)
        {
          case "one":
            result += 1;
            break;
          case "two":
            result += 2;
            break;
          case "three":
            result += 3;
            break;
          case "four":
            result += 4;
            break;
          case "five":
            result += 5;
            break;
          case "six":
            result += 6;
            break;
          case "seven":
            result += 7;
            break;
          case "eight":
            result += 8;
            break;
          case "nine":
            result += 9;
            break;
          case "ten":
            result += 10;
            break;
          case "eleven":
            result += 11;
            break;
          case "twelve":
            result += 12;
            break;
          case "thirteen":
            result += 13;
            break;
          case "fourteen":
            result += 14;
            break;
          case "fifteen":
            result += 15;
            break;
          case "sixteen":
            result += 16;
            break;
          case "seventeen":
            result += 17;
            break;
          case "eighteen":
            result += 18;
            break;
          case "nineteen":
            result += 19;
            break;
          case "twenty":
            result += 20;
            break;
          case "thirty":
            result += 30;
            break;
          case "forty":
            result += 40;
            break;
          case "fifty":
            result += 50;
            break;
          case "sixty":
            result += 60;
            break;
          case "seventy":
            result += 70;
            break;
          case "eighty":
            result += 80;
            break;
          case "ninety":
            result += 90;
            break;
          case "hundred":
            result *= 100;
            break;
          case "thousand":
            thousands = result;
            result = 0;
            break;
          case "million":
            millions = result;
            result = 0;
            break;
        }
      }
      return result + 1000 * thousands + 1000000 * millions;
    }
}